### PR TITLE
[qat] Conv module

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#22899 [qat] Conv module**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D16274792/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #22732 [qat] Quantization aware training in eager mode&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D16199356/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #22714 [qat] Add nn.qat.Linear&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D16193585/)

Added Conv module for qat

Differential Revision: [D16274792](https://our.internmc.facebook.com/intern/diff/D16274792/)

### CODE REVIEW

File: torch/nn/qat/modules/conv.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    # TODO: support initializing from qconfig
+    @staticmethod
+    def from_float(mod):
+        r"""Create a qat module from a float module or qparams_dict
+
+            Args: `mod` a float module, either produced by torch.quantization utilities
+            or directly from user
+        """
+        assert type(mod) == NNConv2d, 'nnq.Linear.from_float only works for nn.Linear'
```
Comment (Lucaskabela) [inline torch/nn/qat/modules/conv.py]: Message should be referring to Conv2d instead of Linear modules

File: torch/nn/quantized/modules/conv.py [hunk 2]
```diff
@@ -108,3 +102,38 @@ def forward(self, input):
...
+
+            Args: `mod` a float module, either produced by torch.quantization utilities
+            or directly from user
+        """
+        if hasattr(mod, 'weight_fake_quant'):
+            # assert type(mod) == QATLinear, 'training mode nnq.Linear.from_float only works for nn.qat.Linear'
+            activation_observer = mod.observer
+            weight_observer = mod.weight_fake_quant
+        else:
+            assert type(mod) == NNLinear, 'nnq.Linear.from_float only works for nn.Linear'
```
Comment (Lucaskabela) [inline torch/nn/quantized/modules/conv.py]: Should this be `type(mod) == NNConv2d` instead?

File: torch/nn/qat/modules/conv.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+from torch.nn.modules.conv import Conv2d as NNConv2d
```
Comment (dzhulgakov) [inline torch/nn/qat/modules/conv.py]: you can just import it from torch.nn directly I think

File: torch/nn/qat/modules/conv.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    """
+
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, dilation=1, groups=1,
+                 bias=True, padding_mode='zeros'):
+        super(Conv2d, self).__init__(in_channels, out_channels, kernel_size,
+                                     stride=stride, padding=padding, dilation=dilation,
+                                     groups=groups, bias=bias, padding_mode=padding_mode)
+        # fake quant module for output activation
+        self.observer = default_qat_qconfig.activation()
```
Comment (dzhulgakov) [inline torch/nn/qat/modules/conv.py]: nit - you can just make it an argument - then you don't need to do reassignment in from_float

File: torch/nn/quantized/modules/linear.py [hunk 5]
```diff
@@ -121,13 +121,36 @@ def weight(self):
     def weight(self, w):
         self._packed_weight = torch.ops.quantized.fbgemm_linear_prepack(w)
 
+    # TODO: These will be duplicated multiple times, should we use multi-inheritance?
```
Comment (dzhulgakov) [inline torch/nn/quantized/modules/linear.py]: why do we need this stuff as properties in the first place?

File: torch/nn/qat/modules/conv.py (hunk 4 above)
Comment (jerryzh168) [reply torch/nn/qat/modules/conv.py]: hmm, is there any value in keeping the API consistent with float conv module?

Comment (raghuramank100) [reply torch/nn/qat/modules/conv.py]: Can we just call self.observer => self.fake_quant? I assume this is related to your poll, but having two different names (observer and fake_quant) and then using if /else or hasattr to determine what is present and then doing the appropriate action seems simpler to me.

File: torch/nn/quantized/modules/linear.py (hunk 5 above)
Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: do we want to enable users to interact with these modules?

Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: cc @zafartahirov

File: torch/nn/qat/modules/conv.py (hunk 4 above)
Comment (jerryzh168) [reply torch/nn/qat/modules/conv.py]: why having two names is simpler? I think worst case we can do self.observer_or_fake_quant and we don't need any of the hasattr checks.

File: torch/nn/quantized/modules/linear.py (hunk 5 above)
Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: this is copied from quantized/modules/conv.py btw.

File: torch/nn/qat/modules/conv.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        self.weight_fake_quant = default_qat_qconfig.weight()
+
+    def forward(self, input):
+        if self.padding_mode == 'circular':
```
Comment (raghuramank100) [inline torch/nn/qat/modules/conv.py]: We have code dup with the nn.Conv2d module here. An option that can save some code will be to write the body of this call as a function with weight as an argument. nn.conv2d would call:
 compute_conv(self.weight,...) while qat.conv would call: compute_conv(self.weight_fake_quant(self.weight)). This requires changing nn.conv2d code.

File: torch/nn/qat/modules/conv.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        assert type(mod) == NNConv2d, 'nnq.Linear.from_float only works for nn.Linear'
+        assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
+        assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
+        qat_conv = Conv2d(mod.in_channels, mod.out_channels, mod.kernel_size,
+                          stride=mod.stride, padding=mod.padding, dilation=mod.dilation,
+                          groups=mod.groups, bias=mod.bias is not None,
+                          padding_mode=mod.padding_mode)
+        qat_conv.weight = mod.weight
+        qat_conv.bias = mod.bias
+        qat_conv.observer = mod.qconfig.activation()
```
Comment (raghuramank100) [inline torch/nn/qat/modules/conv.py]: Similar comment: I assume qat.conv.observer is really a fake_quant and so is weight_fake_quant. Let us have explicit names.

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments, thanks

File: torch/nn/quantized/modules/linear.py (hunk 5 above)
Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: I think we probably don't need to allow user interaction with these things, I'll remove.

File: torch/nn/qat/modules/conv.py (hunk 7 above)
Comment (jerryzh168) [reply torch/nn/qat/modules/conv.py]: planning to use observer for now.

File: torch/nn/qat/modules/conv.py (hunk 4 above)
Comment (jerryzh168) [reply torch/nn/qat/modules/conv.py]: moved to constructor

File: torch/nn/modules/conv.py [hunk 8]
```diff
@@ -199,6 +199,15 @@ def forward(self, input):
         return F.conv1d(input, self.weight, self.bias, self.stride,
                         self.padding, self.dilation, self.groups)
 
+def conv2d_forwad(input, padding_mode, padding, weight, bias, stride, dilation, groups):
```
Comment (Lucaskabela) [inline torch/nn/modules/conv.py]: nit: typo -> should be conv2d_forward

File: torch/nn/modules/conv.py [hunk 9]
```diff
@@ -330,15 +339,8 @@ def __init__(self, in_channels, out_channels, kernel_size, stride=1,
...
-        if self.padding_mode == 'circular':
-            expanded_padding = ((self.padding[1] + 1) // 2, self.padding[1] // 2,
-                                (self.padding[0] + 1) // 2, self.padding[0] // 2)
-            return F.conv2d(F.pad(input, expanded_padding, mode='circular'),
-                            self.weight, self.bias, self.stride,
-                            _pair(0), self.dilation, self.groups)
-        return F.conv2d(input, self.weight, self.bias, self.stride,
-                        self.padding, self.dilation, self.groups)
-
+        conv2d_forwad(input, self.padding_mode, self.padding, self.weight,
```
Comment (Lucaskabela) [inline torch/nn/modules/conv.py]: nit: typo -> conv2d_forward (see https://github.com/pytorch/pytorch/pull/22899/files#r305143145)

File: torch/nn/qat/modules/conv.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+from torch.nn import Conv2d as NNConv2d
+from torch.nn.modules.conv import conv2d_forwad
```
Comment (Lucaskabela) [inline torch/nn/qat/modules/conv.py]: nit: see https://github.com/pytorch/pytorch/pull/22899/files#r305143145

File: torch/nn/qat/modules/conv.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+                 weight_fake_quant=default_qat_qconfig.weight()):
+        super(Conv2d, self).__init__(in_channels, out_channels, kernel_size,
+                                     stride=stride, padding=padding, dilation=dilation,
+                                     groups=groups, bias=bias, padding_mode=padding_mode)
+        # fake quant module for output activation
+        self.observer = activation_fake_quant
+        self.weight_fake_quant = weight_fake_quant
+
+    def forward(self, input):
+        return self.observer(conv2d_forwad(input, self.padding_mode, self.padding,
```
Comment (Lucaskabela) [inline torch/nn/qat/modules/conv.py]: nit: forward typo (https://github.com/pytorch/pytorch/pull/22899/files#r305143145)

File: torch/nn/modules/conv.py [hunk 12]
```diff
@@ -330,15 +339,8 @@ def __init__(self, in_channels, out_channels, kernel_size, stride=1,
...
-        if self.padding_mode == 'circular':
-            expanded_padding = ((self.padding[1] + 1) // 2, self.padding[1] // 2,
-                                (self.padding[0] + 1) // 2, self.padding[0] // 2)
-            return F.conv2d(F.pad(input, expanded_padding, mode='circular'),
-                            self.weight, self.bias, self.stride,
-                            _pair(0), self.dilation, self.groups)
-        return F.conv2d(input, self.weight, self.bias, self.stride,
-                        self.padding, self.dilation, self.groups)
-
+        conv2d_forward(input, self.padding_mode, self.padding, self.weight,
```
Comment (ppwwyyxx) [inline torch/nn/modules/conv.py]: Missing a `return`.
How come there is no tests to catch this?

Comment (jerryzh168) [reply torch/nn/modules/conv.py]: ...will fix. didn't run original conv tests

File: torch/nn/qat/modules/conv.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        >>> m = nn.qat.Conv2d(16, 33, 3, stride=2)
+        >>> # non-square kernels and unequal stride and with padding
+        >>> m = nn.qat.Conv2d(16, 33, (3, 5), stride=(2, 1), padding=(4, 2))
+        >>> # non-square kernels and unequal stride and with padding and dilation
+        >>> m = nn.qat.Conv2d(16, 33, (3, 5), stride=(2, 1), padding=(4, 2), dilation=(3, 1))
+        >>> input = torch.randn(20, 16, 50, 100)
+        >>> output = m(input)
+    """
+
+    __FLOAT_MODULE__
```
Comment (dzhulgakov) [inline torch/nn/qat/modules/conv.py]: dumb question - how does it work?

File: torch/nn/modules/conv.py (hunk 12 above)
Comment (SsnL) [reply torch/nn/modules/conv.py]: There are tests. The merged commit has a red cross shown above. But maybe the open source CI failure signals didn't pass through or something.

Comment (soumith) [reply torch/nn/modules/conv.py]: yea you ignored the tests, i am not sure why

File: torch/nn/qat/modules/conv.py (hunk 13 above)
Comment (jerryzh168) [reply torch/nn/qat/modules/conv.py]: ...need to assign the original module.

**Pull request closed without merging**